Symbia/in context classification of capabilities

leoentersthevoid opened the pull request:
PR for #12 (replaced #13) 

- In-context classification
- Similarity classification (cosine only now)
- minor refactoring for ChatGPT & GPT-4 backends
   - fixed a nasty bug: prompting wasn't part of the user, sometimes it wasn't appended at all 
   - adjusted the system template

run
```bash
pytest tests/test_components.py::TestComponents
```

Xpitfire commented on symai/components.py:
```diff
@@ -301,3 +301,40 @@ def _func(_):
+        self.classes = classes
+        self.metric  = metric
+
+    def embed_classes(self):
```
Should this not use the @cache decorator?

leoentersthevoid replied:
I forgot to tell you in our call. It's getting nasty here. Because cache is a decorator itself, you can't parameterize it from within the class, i.e. change its `in_memory=True/False`. We can't set a default `True` value in here without first telling the user he needs to always be aware of the .cache. I can set its default to `True` for now until we find a solution.

Xpitfire replied:
In this case I would parametrize the class constructor with the in_memory property and set it to False to make the assure a user desired behavior and not a nasty side effect

Xpitfire commented on symai/prompts.py:
```diff
@@ -811,3 +811,52 @@ def __init__(self):
 ])
 
+
+class SymbiaCapabilities(Prompt):
```
You need to also adapt and test this new in-context classification feature with the SymbiaChat, since it changes its behaviour.

Xpitfire commented on symai/components.py:
```diff
@@ -53,27 +56,9 @@ def __init__(self, cases: List[str], default: Optional[str] = None):
         return sym.choice(cases=self.cases, default=self.default, *args, **kwargs)
-    
-
-class Classify(Expression):
```
Did you also trace where the Classify feature was used before removing?

leoentersthevoid replied:
I did, that was the next thing I would have worked on, but the commit will just get too large with changing Symbia as well in here. But sure, I can do it in one go.

Xpitfire replied:
Thank you. I would do it in one go to have a complete major release

Xpitfire reviewed (changes requested):
Please check the comments, and if they are all fulfilled, I will approve the PR.

leoentersthevoid commented:
# Somewhat done
- stable Symbia behavior; refactored the chat.py and fixed things associated with the change
  - in-context tests pass with GPT-4; 
  - added support for 'thought'; we can now take the thoughts resulted from in-context and do ReACT, RCI, etc.
- moved .cache in /home based (handled through `pathlib.Path.home()`) & fixed things associated with the change

`pytest tests/test_components.py::TestComponents::test_symbia_capabilities`

# Next steps
- fix memory; `zip()` it's really bad; need some proper encoding/decoding scheme, because `hash` is a one way mapping and maybe we want the keys as well, not only the values. 
- solving memory should make Symbia workable; now it fails because of the way we upsert into pinecone

leoentersthevoid commented:
- fixed pinecone memory issues; 
- Symbia is operational after the refactor 
- addressed the `zip()`, moved it into `Symbol` because it makes much more sense
- for building the keys, I used `uuid` to generate a random keys for the pinecone; this blends nicely with pinecone because you can add `metadata` in the format `{'text': ...}`, to normally store the query string, which can be fully retrieved at recall time

Xpitfire commented:
Merged substantial fixes
